skbio/tree/_nj.py

```diff
@@ -286,3 +287,332 @@ def _pair_members_to_new_node(dm, i, j, disallow_negative_branch_length):
+    else:
+        root2 = node2.root()
+        nodelist2 = [root2.name]
+        nodelist2.extend(root2.subset() - node2.subset())
+    df = dm.between(nodelist1, nodelist2)
+    return df["value"].mean()
+
+
+def _subtree_count(subtree):
+    """Returns the number of leaves in a subtree.
+
+    Assumes the root as a leaf node.
+    """
+    if subtree.is_tip() or subtree.is_root():
+        return 1
+    else:
+        return subtree.count(tips=True)
+
+
+def _swap_length(a, b, c, d, i, j, k, m, adm):
+    """Returns the change in overall tree length after a given swap.
+
+    The count of leaves contained in each subtree are denoted 'a, b, c, d' while
+    each node defining the subtree has the index 'i, j, k, m', respectively.
+    """
+    lambda1 = (a * d + b * c) / ((a + b) * (c + d))
+    lambda2 = (a * d + b * c) / ((a + c) * (b + d))
+    return 0.5 * (
+        (lambda1 - 1) * (adm[i][k] + adm[j][m])
+        - (lambda2 - 1) * (adm[i][j] + adm[k][m])
+        - (lambda1 - lambda2) * (adm[i][m] + adm[j][k])
+    )
+
+
+def _swap_heap(tree, adm):
+    """Returns a maxheap ordered by the swap length for all possible swaps."""
+    heap = []
+    ordered = list(tree.postorder(include_self=False))
+    root = tree.root()
+    n_taxa = root.count(tips=True) + 1
+    # begin by finding nodes which are the child node of an internal edge
+    for node in ordered:
+        # ignore tips of the tree
+        if node.is_tip():
+            continue
+        # identify the parent and grandparent nodes
+        parent = node
+        a = parent.parent
+        # identify the index of each neighboring node
+        for index, node in enumerate(ordered):
+            if node == a:
+                i1 = index
+        for child in parent.children:
+            if child.is_tip():
+                continue
+            else:
+                childnode = child
+                c, d = childnode.children
+                for sibling in childnode.siblings():
+                    b = sibling
+                for index, node in enumerate(ordered):
+                    if node == b:
+                        i2 = index
+                    elif node == c:
+                        i3 = index
+                    elif node == d:
+                        i4 = index
+                # count the tips of the subtrees defined by the neighboring nodes
+                sub_tips = []
+                for subtree in [b, c, d]:
+                    sub_tips.append(1 if subtree.is_tip() else subtree.count(tips=True))
+                b_, c_, d_ = sub_tips
+                a_ = n_taxa - b_ - c_ - d_
+                # calculate the swap length for the two possible swaps given the edge
+                swap_1 = _swap_length(a_, b_, c_, d_, i1, i2, i3, i4, adm)
+                swap_2 = _swap_length(a_, b_, d_, c_, i1, i2, i4, i3, adm)
+                # store the best possible swap into a maxheap
+                if swap_1 > swap_2 and swap_1 > 0:
```

Comment:
These two conditions are not mutually exclusive. Could there be other conditions?